include/zpp/mutex.hpp

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2019 Erwin Rol <erwin@erwinrol.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR__INCLUDE__ZPP__MUTEX_HPP
+#define ZEPHYR__INCLUDE__ZPP__MUTEX_HPP
+
+#include <zpp/compiler.hpp>
+
+#include <kernel.h>
+#include <sys/__assert.h>
+
+#include <chrono>
+#include <mutex>
+
+namespace zpp {
+
+/**
+ * @brief A recursive mutex class.
+ */
+class mutex {
```

Comment:
I'm not seeing a real value in this or the other owning-wrapper APIs.  Zephyr's mutex and semaphore implementations and API are fine as-is and using them avoids complications of sharing instances with C code.

What we could use is RAII-type classes to wrap them.  It would be great to be able to do something like:
```
if (auto lock = zephyr::take_mutex(mutex, K_MSEC(50))) {
   // do stuff with mutex, automatically freed on completion of block
} else {
   // failed to get mutex; examine lock.status to find out why
}
```

Reply:
Hey @pabigot thanks for your comments. With the RAII and later wrappers you kind of lost me, because isn't your proposal here not exactly the opposite of RAII? When I understand you correct you just want a paper thin wrapper and not really any C++ API ? In the example here I assume `mutex` is a `struct k_mutex*`  ? Which isn't RAII in my eyes. Or am I misunderstanding what you mean here ?

Reply:
`mutex` here is a `struct mutex` instance.  The concept is you'd have something like:
```
template <class Lock>
class take {
      Lock& lock_;
    public:
      take (Lock& lock, k_timeout_t timeout) : lock_{lock} { ... }
}
```
with magic in the constructor to invoke the appropriate operation to take the lock (which could be a mutex, semaphore, spinlock, etc.)

The instance of `take<Lock>` would hold the lock until its scope expired or an unlock method was called directly.  RAII.

Reply:
My `mutex` class is your `Lock`  template argument, and your `take` class is `std::lock_guard`  which I added an alias `zpp::lock_guard` for. 

And am I understanding you want something like;
```
struct k_mutex lock;

void foo() {
{
  take<struct k_mutex> guard(lock, 10);

}
```
And for the rest just use the native C API like `k_mutex_lock(&lock, 0);` ?